edgehog-device-runtime-containers/src/docker/image.rs

```diff
@@ -0,0 +1,680 @@
+// This file is part of Edgehog.
+//
+// Copyright 2023-2024 SECO Mind Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Handle the calls to manage an Image.
+
+use std::{
+    fmt::{Debug, Display},
+    hash::Hash,
+};
+
+use base64::Engine;
+use bollard::{
+    auth::DockerCredentials,
+    errors::Error as BollardError,
+    image::CreateImageOptions,
+    models::{ImageDeleteResponseItem, ImageInspect},
+};
+use futures::{future, TryStreamExt};
+use tracing::{debug, error, info, instrument, trace, warn};
+
+use crate::client::*;
+
+/// Error for the image operations.
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+pub enum ImageError {
+    /// couldn't pull the image
+    Pull(#[source] BollardError),
+    /// couldn't inspect the image
+    Inspect(#[source] BollardError),
+    /// couldn't remove image
+    Remove(#[source] BollardError),
+    /// couldn't remove image in use by container
+    ImageInUse(#[source] BollardError),
+    /// couldn't list the images
+    List(#[source] BollardError),
+    /// couldn't convert container summary
+    Convert(#[from] ConversionError),
+    /// couldn't decode the registry auth
+    RegistryAuth(#[from] RegistryAuthError),
+}
+
+/// Error while decoding the Registry Authentication.
+///
+/// It should be a url safe base64 encoded JSON with padding
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+pub enum RegistryAuthError {
+    /// couldn't base64 decode the registry auth
+    Base64(#[from] base64::DecodeError),
+    /// couldn't decode the registry auth JSON
+    Json(#[from] serde_json::Error),
+}
+
+/// Error for the converting a value to an [`Image`]
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+pub enum ConversionError {
+    /// couldn't get tag from untagged image
+    UntaggedImage,
+    /// couldn't extract the tag for the image
+    MissingTag,
+}
+
+/// Docker image struct.
+#[derive(Debug, Clone, Eq, Hash)]
+pub struct Image<S> {
+    /// Id of the crated image.
+    pub id: Option<String>,
+    /// Reference to the image that we need to pull.
+    pub reference: S,
```

Comment:
I find it generally easier this way, so I don't have to specify a lifetime on the struct for the Cow<'a>.

This way I can easily use the `&'static str, &str, String`, as needed. In the application, this will be a String; this way we don't incur in the overhead of a Cow. On the other hand, in the tests we can construct the Image from a `&'static str`